service/history/ndc/workflow_resetter.go

```diff
@@ -747,6 +756,9 @@ func reapplyEvents(
 				continue
 			}
 			attr := event.GetWorkflowExecutionUpdateAcceptedEventAttributes()
+			if updateRegistry.Contains(attr.ProtocolInstanceId) {
+				continue
+			}
```

Comment:
This is the actual fix.

I considered using `isIncomplete` but that includes `provisionallyIncomplete` which could be rolled back, so `Contains` seemed like safer semantics. However, in theory this could cause us to not reapply an update that is complete but hasn't been removed from the registry. An alternative would be to alter the `isIncomplete` implementation to define `provisionallyComplete` to be incomplete.